### PR TITLE
Fixes #1896: Thoroughly test ItemSelectionInputContainsAtLeastOneOfRuleClassifier 

### LINKED ISSUE #1896: Thoroughly test ItemSelectionInputContainsAtLeastOneOfRuleClassifier
Since the ItemsSelectionInputContainsAtLeastOneOfRuleClassifier was copied from Oppia web, it didn't have corresponding tests added to verify correctness. These should be ported over from Oppia web as well to ensure that each classifier has working functionality.

Useful reference - https://github.com/oppia/oppia-android/tree/develop/domain/src/test/java/org/oppia/android/domain/classify/rules

### PR DESCRIPTION
**Pull request by navneetsaluja**
## Explanation
Fixes #1896 

Tests for lowercase and uppercase inputs. (More test suggestions would be appreciated, this is just to ensure my skeleton is fine.)
(Any help on how to test that these test files are working properly would be appreciated as well)

## Proof
Selection_018

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by navneetsaluja** [comment]
Will be fixing the lint test errors, could I get some leads on how to fix the Roboelectric tests? They are giving the following outputs: 
```
Execution failed for task ':app:mergeDebugJavaResource'.
> Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
   > Could not download byte-buddy.jar (net.bytebuddy:byte-buddy:1.6.11)
      > Could not get resource 'https://jcenter.bintray.com/net/bytebuddy/byte-buddy/1.6.11/byte-buddy-1.6.11.jar'.
         > Could not GET 'https://d29vzk4ow07wi7.cloudfront.net/(long URL)'
            > Connection reset
```
```
> Task :app:kaptGenerateStubsDebugKotlin FAILED
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:kaptGenerateStubsDebugKotlin'.
> Could not resolve all artifacts for configuration ':app:debugCompileClasspath'.
   > Could not download mockito-core.jar (org.mockito:mockito-core:2.7.22)
      > Could not get resource 'https://jcenter.bintray.com/org/mockito/mockito-core/2.7.22/mockito-core-2.7.22.jar'.
         > Could not GET '(LONG URL)'
            > Connection reset
```

**Comment by anandwana001** [comment]
@ARJUPTA @prayutsu would you like to help the contributor here with few initial thoughts on this.

@navneetsaluja Check our wiki guide here for ktlint issues - https://github.com/oppia/oppia-android/wiki/Ktlint-Guide

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (Arjupta) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Max 100 characters should be there in one.  Keep your code behind the grey line of Android Studio.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (Arjupta) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Same maximum 100 character exceeded error

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (Arjupta) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Change the order of your imports in this order without space.
"java"
"javax"
"kotlin" 
aliases in the end

**Review by Arjupta — CHANGES_REQUESTED** [review]
This will get your failing lint checks done

**Comment by Arjupta** [comment]
> Will be fixing the lint test errors, could I get some leads on how to fix the Roboelectric tests? They are giving the following outputs:
> 
> ```
> Execution failed for task ':app:mergeDebugJavaResource'.
> > Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
>    > Could not download byte-buddy.jar (net.bytebuddy:byte-buddy:1.6.11)
>       > Could not get resource 'https://jcenter.bintray.com/net/bytebuddy/byte-buddy/1.6.11/byte-buddy-1.6.11.jar'.
>          > Could not GET 'https://d29vzk4ow07wi7.cloudfront.net/(long URL)'
>             > Connection reset
> ```
> 
> ```
> > Task :app:kaptGenerateStubsDebugKotlin FAILED
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':app:kaptGenerateStubsDebugKotlin'.
> > Could not resolve all artifacts for configuration ':app:debugCompileClasspath'.
>    > Could not download mockito-core.jar (org.mockito:mockito-core:2.7.22)
>       > Could not get resource 'https://jcenter.bintray.com/org/mockito/mockito-core/2.7.22/mockito-core-2.7.22.jar'.
>          > Could not GET '(LONG URL)'
>             > Connection reset
> ```

Have you tried this after connecting to internet?

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Yes, already fixed that. Will push in the afternoon. Thank you.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: I put .builder() on another line, but the name of `DaggerItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest_TestApplicationComponent` is 100 characters itself, will it work?

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Alright, thank you.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (Arjupta) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Yes, it just have to be maximum 100

**Comment by navneetsaluja** [comment]
@anandwana001 Thank you!

@ARJUPTA 
> Have you tried this after connecting to internet? 

I'm not sure what you mean. All these tests were done on Github while pushing ([here](https://github.com/oppia/oppia-android/pull/2073/checks?check_run_id=1343685925)). Other than that, my local instance is always connected to the internet.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Thank you!

**Comment by Arjupta** [comment]
> @anandwana001 Thank you!
> 
> @ARJUPTA
> 
> > Have you tried this after connecting to internet?
> 
> I'm not sure what you mean. All these tests were done on Github while pushing ([here](https://github.com/oppia/oppia-android/pull/2073/checks?check_run_id=1343685925)). Other than that, my local instance is always connected to the internet.

I think you should run these tests once again. So just make a commit after correcting the lint error, it will automatically trigger the tests

**Comment by prayutsu** [comment]
Hi @navneetsaluja 
I think this classifier is not about `uppercase` and `lowercase` strings. You should take a look at `ItemSelectionContainsAtLeastOneOfRule` and then try figuring out what you need to do.
[ItemSelectionContainsAtLeastOneOfRule](https://github.com/oppia/oppia-android/blob/develop/domain/src/main/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProvider.kt)
Also, remove the `TODO` in the file `ItemSelectionContainsAtLeastOneOfRule`

**Comment by prayutsu** [comment]
> Will be fixing the lint test errors, could I get some leads on how to fix the Roboelectric tests? They are giving the following outputs:
> 
> ```
> Execution failed for task ':app:mergeDebugJavaResource'.
> > Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
>    > Could not download byte-buddy.jar (net.bytebuddy:byte-buddy:1.6.11)
>       > Could not get resource 'https://jcenter.bintray.com/net/bytebuddy/byte-buddy/1.6.11/byte-buddy-1.6.11.jar'.
>          > Could not GET 'https://d29vzk4ow07wi7.cloudfront.net/(long URL)'
>             > Connection reset
> ```
> 
> ```
> > Task :app:kaptGenerateStubsDebugKotlin FAILED
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':app:kaptGenerateStubsDebugKotlin'.
> > Could not resolve all artifacts for configuration ':app:debugCompileClasspath'.
>    > Could not download mockito-core.jar (org.mockito:mockito-core:2.7.22)
>       > Could not get resource 'https://jcenter.bintray.com/org/mockito/mockito-core/2.7.22/mockito-core-2.7.22.jar'.
>          > Could not GET '(LONG URL)'
>             > Connection reset
> ```

These tests are giving this error because these are not relevant to this particular classifier, in this test we want to test if the `inputs` and `answer` have any common item or not. And that's what the name suggests, "contains at least one" (answer should contain at least one of the item in inputs).

**Comment by prayutsu** [comment]
@navneetsaluja You can do one more thing, you can add keywords `Fixes #bugnum` in the explanation section itself and not above the explanation section, then this would be more close to our conventional PR description

**Comment by navneetsaluja** [comment]
@ARJUPTA Will commit again once I implement prayutsu's suggestions. Thank you.

@prayutsu I see, I already have three tests in mind, will try implementing them. Will edit the description of the PR as well to suit the conventions. Thank you for your input!

**Comment by rt4914** [comment]
Thanks a lot @prayutsu and @ARJUPTA for helping out new contributors. Thanks a lot.

**Comment by navneetsaluja** [comment]
@prayutsu @ARJUPTA PTAL, checks passed, thanks. 
Updated the tests too: they use a common Item Set as answer which is 
`["test1","test2","test3","test4","test5"]` 
I feel that the variables and the functions have bad names - not sure what to name them. Any more ideas for tests would also be helpful. Thanks!

**Comment by Arjupta** [comment]
> @prayutsu @ARJUPTA PTAL, checks passed, thanks.
> Updated the tests too: they use a common Item Set as answer which is
> `["test1","test2","test3","test4","test5"]`
> I feel that the variables and the functions have bad names - not sure what to name them. Any more ideas for tests would also be helpful. Thanks!

Changes I mentioned are done correctly. LGTM Thanks @navneetsaluja

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (prayutsu) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt:65]: @navneetsaluja It would be great if we add some cases in which more than one item is common.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (prayutsu) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt:78]: The empty set is a subset of all the sets so we can also add some test cases in which input is an empty set.
@rt4914 What do you think about this?

**Review by prayutsu — CHANGES_REQUESTED** [review]

**Comment by prayutsu** [comment]
@navneetsaluja I think variable names are fine, but I am not sure about the test case names, maybe @anandwana001 can suggest some good name.

**Comment by prayutsu** [comment]
@navneetsaluja It would be more close to our conventional title for the PRs if you add a colon after `Fixes #bugnum` like this `Fixes #1896:` 
Also, it's good to have the PR title the same as the issue name.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt:65]: @prayutsu Alright, will add cases with a set with 2 common items and one more with 2 common items and one (and more) not common item(s).

**Comment by navneetsaluja** [comment]
@prayutsu I see, will take care of it in future PRs. Thanks.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (rt4914) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt:78]: @prayutsu Yes I do think it would be a nice idea.

**Comment by navneetsaluja** [comment]
@prayutsu new commit, PTAL :)

**Review by prayutsu — APPROVED** [review]
@navneetsaluja Great work!
 LGTM, thanks!

**Review by Arjupta — APPROVED** [review]
Thanks @navneetsaluja

**Comment by anandwana001** [comment]
@ARJUPTA @prayutsu please resolve your comments, if you think the required changes are made.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: No empty line required here

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Leave an empty above

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Use `InteractionObjectTestBuilder` to call functions like `createStringList`.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Thanks @navneetsaluja,  added a nit change. 
All test cases are passing for me. 

Screenshot 2020-11-09 at 08 30 12


Update your branch with the latest develop, currently, your branch is 30 commits behind the latest develop.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Do you mean using the already defined `createHtmlStringList` function from `InteractionObjectTestBuilder`?

**Review by Arjupta — APPROVED** [review]

**Comment by Arjupta** [comment]
> @ARJUPTA @prayutsu please resolve your comments, if you think the required changes are made.

@anandwana001 i think only the author of PR and people with write access can resolve them, as I am unable to find the resolve conversation button

**Comment by prayutsu** [comment]
> @ARJUPTA @prayutsu please resolve your comments, if you think the required changes are made.

Actually, I am not getting the option to resolve comments.

**Comment by anandwana001** [comment]
@ARJUPTA @prayutsu oh, ok, I got it. Thanks for update. The author will resolve it.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: @anandwana001 forgot to ping you before, check the message above please ^

Comment (anandwana001) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: @navneetsaluja yes, all the functions are inside `InteractionObjectTestBuilder `. We need to use those and remove which are in the test files.

Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: @anandwana001 Got it, will push changes at 7 after classes.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Removed.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Added, should be okay now.

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (anandwana001) [inline domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: shift this method to `InteractionObjectTestBuilder`

Comment (anandwana001) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Also, check if there is already any method in `InteractionObjectTestBuilder ` class and if we can use them directly.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Thanks @navneetsaluja 
nit change added

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (Sarthak2601) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: I'd suggest direct use of `createHtmlStringList` here if it's possible.

Comment (navneetsaluja) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: @Sarthak2601 already used that, have I used it incorrectly? I tried direct StringLists but IIRC either `mapOf` or `assertThat` functions require `InteractionObject`s (will confirm which one after 4). All tests require `InteractionObject`'s so I was thinking instead of using a custom function to change a `StringList` to `InteractionObject`, maybe use the already defined `createListOfSetsOfHtmlStrings` and then do it.

**Comment by navneetsaluja** [comment]
@/anandwana001(Not pinging because you've already been assigned)
 Sorry for the late update, had real life commitments. Was thinking of using the `createListOfSetsOfHtmlStrings` defined in the `InteractionObjectTestBuilder.kt` but then it was rendering code a bit unreadable, so I just shifted my function to `InteractionObjectTestBuilder.kt`. 
Other than that, it gives an App module test error here:
`Caused by: org.gradle.process.internal.ExecException: Process 'Gradle Test Executor 15' finished with non-zero exit value 134
This problem might be caused by incorrect test process configuration.`
; but on my local machine all the tests passed. 

UPDATE: Checks passed now, I didn't change anything. When I wrote the original comment it was displaying the error above.

File: domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt
Comment (Sarthak2601) [reply domain/src/test/java/org/oppia/android/domain/classify/rules/itemselectioninput/ItemSelectionInputContainsAtLeastOneOfRuleClassifierProviderTest.kt]: Yes, that's okay but I think you can shift this function -- `createStringListInteractionObject` to the `InteractionObjectTestBuilder` as all the interaction objects for test cases are getting handled from there.

**Review by Sarthak2601 — APPROVED** [review]
LGTM, thanks!

**Review by anandwana001 — APPROVED** [review]
LGTM, Thanks @navneetsaluja 
Before the merge, could you please run all the test cases and put a screenshot in the description showing all tests are passing?

@rt4914 Adding you for further merge.

**Comment by navneetsaluja** [comment]
@anandwana001 Added screenshot!

**Pull request merged**
